### PR TITLE
[cxx-interop] Use VFS to inject modulemap into libstdc++ installation

### PR DESCRIPTION
**Pull request by egorzhdan**
Previously the modulemap for the C++ stdlib on Linux was provided via `-fmodule-map-file=` Clang argument pointing to the modulemap file within the Swift toolchain. The modulemap file could not reference the stdlib headers directly, since the exact stdlib include directory varies across Linux versions (it generally looks like `/usr/include/c++/{gcc_version}`). So the modulemap file instead referenced a local header, which `#include <>`-ed the stdlib headers, relying on Clang include resolution.

Unfortunately this did not work properly in the presence of another C++ module which included the stdlib headers: sometimes decls from the stdlib were hijacked by the other module, and were not treated as a part of the stdlib by Clang. This caused compile errors in Swift.

This change uses LLVM VFS to inject the modulemap file into the libstdc++ directory. The modulemap file is now able to reference the stdlib headers directly, which fixes the issue.

Credits to Rintaro Ishizaki for proposing a similar idea for SwiftGlibc back in 2016.

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
Please test with following PR:
https://github.com/apple/llvm-project/pull/4675

@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@rintaro this is largely based on your commit back from 2016 that never got merged (https://github.com/rintaro/swift/commit/1dc920bc971f29897155873086ac0fe2292705be).
Thanks for proposing this idea and working on it!

**Comment by egorzhdan** [comment]
This is a requirement for https://github.com/apple/swift/pull/58699.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by rintaro** [comment]
Thank you for resurrecting this idea :)

File: lib/ClangImporter/ClangImporter.cpp [hunk 1]
```diff
@@ -870,6 +863,92 @@ importer::addCommonInvocationArguments(
...
+  using Path = SmallString<128>;
+
+  const llvm::Triple &triple = ctx.LangOpts.Target;
+  // We currently only need this when building for Linux.
+  if (!triple.isOSLinux())
+    return {};
+
+  SearchPathOptions &searchPathOpts = ctx.SearchPathOpts;
+
+  Path SDKPath(searchPathOpts.getSDKPath());
```
Comment (zoecarver) [inline lib/ClangImporter/ClangImporter.cpp:880]: Nit: camelCase

File: stdlib/public/Cxx/libstdcxx.h [hunk 2]
```diff
@@ -1,79 +1,79 @@
-#include <algorithm>
```
Comment (zoecarver) [inline stdlib/public/Cxx/libstdcxx.h:1]: Why did these have to change? Is it ever going to be a problem for users if they use angle brackets in their includes?

**Review by zoecarver — APPROVED** [review]
Looks good to me. Thanks, Egor. Ship it when you feel ready!

File: stdlib/public/Cxx/libstdcxx.h (hunk 2 above)
Comment (egorzhdan) [reply stdlib/public/Cxx/libstdcxx.h:1]: It'll work fine with angle brackets as well. I just wanted to make it more clear that both the modulemap & the header are now located in the same directory as the stdlib headers, so there's no need to use angle brackets.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by aschwaighofer** [comment]
Hi @egorzhdan,

Can you fix the line 

```
/home/build-user/swift/lib/ClangImporter/ClangImporter.cpp:869:8: error: too few template arguments for class template 'SmallVector'
static SmallVector<std::pair<std::string, std::string>>
       ^
/home/build-user/swift/include/swift/Basic/LLVM.h:47:43: note: template is declared here
  template <typename T, unsigned N> class SmallVector;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       ^
/home/build-user/swift/lib/ClangImporter/ClangImporter.cpp:926:3: error: too few template arguments for class template 'SmallVector'
  SmallVector<std::pair<std::string, std::string>> result;
  ^
/home/build-user/swift/include/swift/Basic/LLVM.h:47:43: note: template is declared here
  template <typename T, unsigned N> class SmallVector;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~       ^
2 errors generated.
```

SmallVector without a on stack storage template parameter causes compile failure on some linux configurations:

https://ci.swift.org/view/Dashboard/job/oss-swift-package-centos-7/480/

File: lib/ClangImporter/ClangImporter.cpp [hunk 3]
```diff
@@ -870,6 +863,92 @@ importer::addCommonInvocationArguments(
...
+  std::error_code errorCode;
+  for (llvm::vfs::directory_iterator
+           iter = ctx.SourceMgr.getFileSystem()->dir_begin(cxxStdlibsRoot,
+                                                           errorCode),
+           endIter;
+       !errorCode && iter != endIter; iter = iter.increment(errorCode)) {
+    cxxStdlibDirs.push_back(Path(iter->path()));
+  }
+
+  SmallVector<std::pair<std::string, std::string>> result;
```
Comment (aschwaighofer) [inline lib/ClangImporter/ClangImporter.cpp:926]: ^ Here

**Comment by aschwaighofer** [comment]
@egorzhdan 
I put up a fix here: https://github.com/apple/swift/pull/58922

**Pull request merged**